### ISSUE TITLE
failed at the last step

### ISSUE THREAD

**Issue by cbiitbian** [bioinform/neusomatic#53]
example sample ran well, but when run with a pair of real data set, neusomatic failed at the last step with error:
### NeuSomatic ensemble FAILED: Files ../example/work_ensemble/NeuSomatic_ensemble.vcf and ../NeuSomatic_ensemble.vcf Are Different! ###
Please advise what could be wrong

**Comment by cbiitbian** [comment]
I see what is wrong here. The last step of comparison is for the testing sample to make sure testing process run successfully. 
Can I substitute the input files with my sample sets including normal bam, tumor bam, bed and reference genome and consider the result vcf at /example/work_standalone/NeuSomatic_standalone.vcf  and /example/work_ensemble/NeuSomatic_ensemble.vcf are the variant results for my data?

**Comment by msahraeian** [comment]
@cbiitbian happy to see your interest in NeuSomatic. As you pointed out the test script will only succeed if it is used on test data. For your real application, you should follow the instructions in README. The inputs/models/parameters to use may need modification depending on your data.
Particularly, 
1- I would recommend to use newer trained models (NeuSomatic_v0.1.4_standalone_SEQC-WGS-Spike.pth for standalone and NeuSomatic_v0.1.4_ensemble_SEQC-WGS-Spike.pth for ensemble), more information [here](https://github.com/bioinform/neusomatic#trained-network-models). Depending how different your data is from our training set you may be interested to train the network too.
2- For ensemble mode, you need to have somatic calls from five other tools (MuTect2, VarDict, Strelka2, SomaticSniper, and MuSE) for your samples, you can find instruction for that [here](https://github.com/bioinform/neusomatic#ensemble-mode)
3- The reference/regions/minimum AFs can also be adjusted as explained [https://github.com/bioinform/neusomatic#example-usage](url). As discussed there, for inference, you need to follow three steps (preprocess/call/postprocess).

Please let me know if you need more information

**Comment by cbiitbian** [comment]
Hi, Msahraeian
Thanks for the quick response.
I followed your instruction and successfully rerun a standalone test of chr1 of my sample data, but I have a couple of questions I’d like to get your help.

  1.  I basically use the same code as the testing script, but substitute the ref genome, bed, tumor bam and normal bam, I use the NeuSomatic_v0.1.4_standalone_SEQC-WGS-Spike.pth as pretrained model and ran the three steps preprocess/call/postprocess, is this the right way?
  2.  Is “work_standalone/NeuSomatic_standalone.vcf” the final variant call result? The reason I am asking is because the number of variant (270) is almost  triple that in the grand truth (95), which represent a lot of false positive even before a comparison. Maybe the parameters need to adjust or the pretrained model is not a good match?
  3.  I saw in the paper, you use cll sample and got very good results. I al so have the sample and grand truth, I’d like to follow your protocol to do it, I also have the DREAM sets mentioned in your paper. Would you let me know the parameters and trained model you use so that I can re-run?
  4.  If it is necessary, I’d like to learn bamsuegeon and produce my own trained model, so any information about how you did spike in would also be greatly appreciated.

Thanks very much!
Xiaopeng Bian
CBIIT/NCI

**Comment by msahraeian** [comment]
@cbiitbian 
1. Yes, it should be the right steps to go. You may also set the number of threads based on your resources. Another important note is that on the test script I used min AF of 5%. If it is too high for your sample, you can modify the parameters. For our new [SEQC-II robustness analysis paper](https://doi.org/10.1101/667261), we use the following setting:  `--scan_maf 0.01 --min_mapq 10 --snp_min_af 0.03 --snp_min_bq 15 --snp_min_ao 3 --ins_min_af 0.02 --del_min_af 0.02`. 
2. Each variant call is assigned a quality score in the VCF. Based on this score we categorized the calls to three classes PASS (score>=0.7), LowQUAL (0.4<=score<0.7), and REJECT (score<0.4). This is the default limits we use. You can modify these limits in the postprocessing step using input parameters. You should then use the PASS calls for analysis. 
3. For real samples in [NatureComm paper](https://doi.org/10.1038/s41467-019-09027-x), we used DREAM challenge model `NeuSomatic_v0.1.3_standalone_Dream3.pth` to infer. There, we used 0.97 pass threshold for CLL1 and COLO-829 samples. For the DREAM challenge stage 3 we also used DREAM challenge model to infer with 0.7 pass threshold. For the [SEQC-II robustness analysis paper](https://doi.org/10.1101/667261), we used the default 0.7 pass threshold for whole analysis. With the [new SEQC trained models](https://github.com/bioinform/neusomatic#trained-network-models), 0.7 pass threshold should give you similar accuracy on CLL1 and COLO samples. 
4. You can find the information on how to use BAMSurgeon to spike in mutations [here](https://github.com/bioinform/neusomatic#creating-training-data).

Also note that the best performance is usually achieved in the ensemble mode.

**Comment by cbiitbian** [comment]
Hi,
Another question, I am running WES (Whole Exome Sequencing) sample, do you have trained model for that or I can use WGS model? I saw you have a older model for WEX, “NeuSomatic_v0.1.0_standalone_WEX_100purity.pth”, is that a WES model I can use?
Thanks.
Xiaopeng

**Comment by msahraeian** [comment]
@cbiitbian The `NeuSomatic_v0.1.0_standalone_WEX_100purity.pth` is the model we used in original paper for WES. We have also applied our WGS trained model `NeuSomatic_v0.1.4_standalone_SEQC-WGS-Spike.pth` on WES data in the SEQC paper and it seems to perform good.

**Comment by cbiitbian** [comment]
Thanks,
I am trying both model based on your information. The process terminated due to error. I could not figure out how to fix it. I am attaching the error log for the run and a log file suggested by the process log:
Please check error log at work_standalone/work_tumor/work.212/scan.err
Could you please have a look and give me some advice on how to fix it?
Thanks very much.
Xiaopeng Bian

**Comment by msahraeian** [comment]
@cbiitbian I cannot see the attachment. Would you please attach the `scan.err` file again?

**Comment by cbiitbian** [comment]
Here is all that in the scan.err file:
terminate called after throwing an instance of 'std::ios_base::failure[abi:cxx11]'
  what():  basic_ios::clear: iostream error

**Comment by msahraeian** [comment]
@cbiitbian it is not clear to me. Can you make sure you alignment `.bam` files have index files `.bam.bai` with them. Also please make sure the reference file used to align reads matches the one you use here. If it didn't help, can you restrict your bam to a smaller region and send it to me (if the code fails on the restricted bam).

**Comment by cbiitbian** [comment]
Hi,
I am testing what you suggested and will let you know.
In the mean time I got another error for another sample, would you please help me out.
ERROR 2019-12-03 05:39:01,890 run_scan_alignments (ForkPoolWorker-14) Please check error log at work_standalone/work_tumor/work.148/scan.err
Thanks.
Xiaopeng

**Comment by cbiitbian** [comment]
Hi,
I did the test you suggested, I used the problematic region bed to restricted the bam to a small region without change anything else, it worked well without any error.
I’ll use the full bed file and try again, hope it is just a random error, but I do see this error in a lot of samples, though some samples ran successfully, they are all very similar samples.
Thanks.
Xiaopeng Bian

**Comment by msahraeian** [comment]
@cbiitbian if it is a public sample, it would be good if you could share with me the whole bam file and the command line parameters you use and I will take a look.

**Comment by cbiitbian** [comment]
I am sorry this is not a public data set.
We are comparing performance of 5 variant caller, including Neusomatic
Most of the samples ran successfully, however, the results have no overlap with the results of other 4 callers which have pretty good overlap among them. I am worried that I may not have run the process correctly? I am attaching the scripts I used, which was basically modified from the example script included in the package. Would you have a look and give some suggestions?
I am attaching both WGS and WEX models scripts.
Thanks very much!
Xiaopeng Bian

**Comment by msahraeian** [comment]
@cbiitbian I think there should be sth wrong with how you run NeuSomatic. I cannot see your attachments. Would you please send them again. You can attach them to the Github message.

**Comment by cbiitbian** [comment]
Hi,
When I compared with results from two other centers, there are good overlaps, so I think there maybe something wrong with the data from the other center. The only thing is that Neusomatic called more variants than other callers. I am attaching my run scripts and a VCFcomparator somparison result for your reference.
My scripts were named with .sh, which maybe blocked by your system last time, so I renamed it to .txt, hope will get through.
Let me know if you still cannot receive them.
Thanks.
Xiaopeng

**Comment by msahraeian** [comment]
@cbiitbian are you using only PASS calls during evaluation? And if your sample is WEX you should intersect the final vcf with the exons bed.

**Comment by cbiitbian** [comment]
Hi,
The Neusomatic  output default to the same folder as the bam file folder. Can I redirect to another directory?
Thanks.
Xiaopeng

**Comment by msahraeian** [comment]
@cbiitbian you don't need to output to the same directory. For each step there is an input parameter that specifies the output location. for `preprocess.py` you can specify the work folder with `--work`, for `call.py` you can specify the output folder that with `--out`, and for `postprocess.py` you can specify the output vcf using `--output_vcf ` and the work folder with `--work`.

**Comment by cbiitbian** [comment]
Hi,
I ran Neusomatic on some Hapmap samples but the results still showed very low overlap with the “ground truth” they provided. These samples were aligned to hg38 genome, does that have to be the same with the trained model? I am very anxious to get some idea why the overlap is so low?
Thanks.
Xiaopeng

**Comment by cbiitbian** [comment]
For hapmap samples, which trained model do you suggest I should use, any particular attention, like the parameters, I should pay to?
Thanks.
Xiaopeng

**Comment by msahraeian** [comment]
@cbiitbian I think there is something wrong here. Are you using HapMap samples as tumor-normal pairs? As HapMap data is public, it would be great if you could share with me the ID of samples you are using as well as your neusomatic run scripts and I can make sure you are using the right settings.

**Comment by cbiitbian** [comment]
Hi,
Sorry, I don’t have the hapmap id of the files because the data were re-processed and renamed. But I found that the “ground truth” provided contains a lot of duplicate after annotation, so I removed the duplicate and re-compared them, the overlap got much better but still low and the false positive is quite high. I attached my run script and one of the vcfcomparator results FYI. Maybe can manipulate the parameters to get better results? Your suggestion would be highly appreciated.
Thanks.
Xiaopeng
